Unescaped command-line arguments for Windows

kornelski opened the pull request:
Some Windows commands, expecially `cmd.exe /c`, have unusual quoting requirements which are incompatible with default rules assumed for `.arg()`.

This adds `.unquoted_arg()` to `Command` via Windows `CommandExt` trait.

Fixes #29494

rust-highfive commented:
r? @yaahc

(rust-highfive has picked a reviewer for you, use r? to override)

kornelski commented:
There's an existing unstable feature `force_quotes()` that is the opposite of this feature — used to quote args that have `*` which aren't quoted by default in `.arg()`. 

I went for `.unquoted_arg()` approach, because it's more flexible. A single `.unquoted_arg()` can be used to build an arbitrary command-line, but also can be mixed with regular `.arg()`, so that users who need only some arguments with special syntax don't have to reinvent their own quoting implementation for the rest (the escaping rules for Windows are very weird, so it's better not to have to do it).

I think `.force_quotes(true)` could be replaced by `.quoted_arg(val)` to have both features with consistent interface, but I'm leaving it out of scope of this PR.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2705208834) [(plain)](https://github.com/rust-lang/rust/commit/87d364a3dfd8d4d5211ad4d1236b29025c5a8104/checks/2705208834/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
    Checking std_detect v0.1.5 (/checkout/library/stdarch/crates/std_detect)
    Checking miniz_oxide v0.4.0
    Checking object v0.22.0
    Checking addr2line v0.14.0
error[E0599]: no method named `unquoted_arg` found for mutable reference `&mut sys::windows::process::Command` in the current scope
   --> library/std/src/os/./windows/process.rs:151:29
    |
151 |         self.as_inner_mut().unquoted_arg(raw_text);
    |                             ^^^^^^^^^^^^ method not found in `&mut sys::windows::process::Command`
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `CommandExt` defines an item `unquoted_arg`, perhaps you need to implement it
   --> library/std/src/os/./windows/process.rs:105:1
    |
105 | pub trait CommandExt: Sealed {

error[E0308]: mismatched types
   --> library/std/src/sys/windows/process.rs:155:23
    |
    |
155 |         CommandArgs { iter }
    |                       ^^^^ expected struct `OsString`, found enum `Arg`
    |
    = note: expected struct `core::slice::Iter<'_, OsString>`
               found struct `core::slice::Iter<'_, Arg>`

error[E0277]: `Arg` doesn't implement `core::fmt::Debug`
   --> library/std/src/sys/windows/process.rs:263:32
263 |               write!(f, " {:?}", arg)?;
    |               -------------------^^^-
    |               |                  |
    |               |                  |
    |               |                  `Arg` cannot be formatted using `{:?}`
    | 
   ::: /checkout/library/core/src/macros/mod.rs:472:1
    |
472 | / macro_rules! write {
472 | / macro_rules! write {
473 | |     ($dst:expr, $($arg:tt)*) => ($dst.write_fmt($crate::format_args!($($arg)*)))
474 | | }
474 | | }
    | |_- in this expansion of `write!` (#1)
829 | /     macro_rules! format_args {
829 | /     macro_rules! format_args {
830 | |         ($fmt:expr) => {{ /* compiler built-in */ }};
831 | |         ($fmt:expr, $($args:tt)*) => {{ /* compiler built-in */ }};
    | |_____- in this expansion of `$crate::format_args!` (#2)
    |
    = help: the trait `core::fmt::Debug` is not implemented for `Arg`
    = help: the trait `core::fmt::Debug` is not implemented for `Arg`
    = note: add `#[derive(Debug)]` or manually implement `core::fmt::Debug`
    = note: required because of the requirements on the impl of `core::fmt::Debug` for `&Arg`
    = note: required by `core::fmt::Debug::fmt`
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0277, E0308, E0599.
For more information about an error, try `rustc --explain E0277`.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `std`

To learn more, run the command again with --verbose.
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "i686-pc-windows-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu --all-targets
Build completed unsuccessfully in 0:00:17
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2705244641) [(plain)](https://github.com/rust-lang/rust/commit/f450b0dbe3b15898af72410f3b9a1d0e6550bd15/checks/2705244641/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
    Checking addr2line v0.14.0
error[E0408]: variable `s` is not bound in all patterns
   --> library/std/src/sys/windows/process.rs:593:30
    |
593 |             Arg::Quoted(s) | Arg::Unquoted => s.as_ref(),
    |                         -    ^^^^^^^^^^^^^ pattern doesn't bind `s`
    |                         variable not in all patterns


error[E0425]: cannot find value `arg` in this scope
   --> library/std/src/sys/windows/process.rs:150:38
    |
150 |         self.args.push(Arg::Unquoted(arg.to_os_string()))


error[E0532]: expected unit struct, unit variant or constant, found tuple variant `Arg::Unquoted`
   --> library/std/src/sys/windows/process.rs:593:30
    |
104 |     Unquoted(OsString),
    |     ------------------ `Arg::Unquoted` defined here
...
593 |             Arg::Quoted(s) | Arg::Unquoted => s.as_ref(),
    |                              ^^^^^^^^^^^^^ help: use the tuple variant pattern syntax instead: `Arg::Unquoted(_)`

error[E0277]: `Arg` doesn't implement `core::fmt::Debug`
   --> library/std/src/sys/windows/process.rs:267:32
267 |               write!(f, " {:?}", arg)?;
    |               -------------------^^^-
    |               |                  |
    |               |                  |
    |               |                  `Arg` cannot be formatted using `{:?}`
    | 
   ::: /checkout/library/core/src/macros/mod.rs:472:1
    |
472 | / macro_rules! write {
472 | / macro_rules! write {
473 | |     ($dst:expr, $($arg:tt)*) => ($dst.write_fmt($crate::format_args!($($arg)*)))
474 | | }
474 | | }
    | |_- in this expansion of `write!` (#1)
829 | /     macro_rules! format_args {
829 | /     macro_rules! format_args {
830 | |         ($fmt:expr) => {{ /* compiler built-in */ }};
831 | |         ($fmt:expr, $($args:tt)*) => {{ /* compiler built-in */ }};
    | |_____- in this expansion of `$crate::format_args!` (#2)
    |
    = help: the trait `core::fmt::Debug` is not implemented for `Arg`
    = help: the trait `core::fmt::Debug` is not implemented for `Arg`
    = note: add `#[derive(Debug)]` or manually implement `core::fmt::Debug`
    = note: required because of the requirements on the impl of `core::fmt::Debug` for `&Arg`
    = note: required by `core::fmt::Debug::fmt`

error[E0277]: `Arg` doesn't implement `core::fmt::Debug`
   --> library/std/src/sys/windows/process.rs:612:32
    |
612 |         f.debug_list().entries(self.iter.clone()).finish()
    |                                ^^^^^^^^^^^^^^^^^ `Arg` cannot be formatted using `{:?}`
    = help: the trait `core::fmt::Debug` is not implemented for `Arg`
    = help: the trait `core::fmt::Debug` is not implemented for `Arg`
    = note: add `#[derive(Debug)]` or manually implement `core::fmt::Debug`
    = note: required because of the requirements on the impl of `core::fmt::Debug` for `&Arg`
error: aborting due to 5 previous errors

Some errors have detailed explanations: E0277, E0408, E0425, E0532.
For more information about an error, try `rustc --explain E0277`.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `std`

To learn more, run the command again with --verbose.
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "i686-pc-windows-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu --all-targets
Build completed unsuccessfully in 0:00:17
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2705313132) [(plain)](https://github.com/rust-lang/rust/commit/0b2e50210e3b53e5dd372d6644b15758dc48a6c9/checks/2705313132/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
    Checking hashbrown v0.11.0
    Checking miniz_oxide v0.4.0
    Checking std_detect v0.1.5 (/checkout/library/stdarch/crates/std_detect)
    Checking addr2line v0.14.0
error[E0425]: cannot find value `arg` in this scope
   --> library/std/src/sys/windows/process.rs:151:38
    |
151 |         self.args.push(Arg::Unquoted(arg.to_os_string()))

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
For more information about this error, try `rustc --explain E0425`.
error: could not compile `std`

To learn more, run the command again with --verbose.
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "i686-pc-windows-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu --all-targets
Build completed unsuccessfully in 0:00:13
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2705564200) [(plain)](https://github.com/rust-lang/rust/commit/942718299017625b9fdcd22a04965167f3cc486d/checks/2705564200/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
    Checking rand v0.7.3
    Checking std v0.0.0 (/checkout/library/std)
    Checking alloc v0.0.0 (/checkout/library/alloc)
    Checking core v0.0.0 (/checkout/library/core)
error[E0277]: a value of type `realstd::vec::Vec<ffi::os_str::OsString>` cannot be built from an iterator over elements of type `Arg`
  --> library/std/src/sys/windows/process/tests.rs:29:66
   |
29 |             &args.iter().map(|a| Arg::Quoted(OsString::from(a))).collect::<Vec<OsString>>(),
   |                                                                  ^^^^^^^ value of type `realstd::vec::Vec<ffi::os_str::OsString>` cannot be built from `std::iter::Iterator<Item=Arg>`
   |
   = help: the trait `FromIterator<Arg>` is not implemented for `realstd::vec::Vec<ffi::os_str::OsString>`
error[E0308]: mismatched types
  --> library/std/src/sys/windows/process/tests.rs:29:13
   |
   |
29 |             &args.iter().map(|a| Arg::Quoted(OsString::from(a))).collect::<Vec<OsString>>(),
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected slice, found struct `realstd::vec::Vec`
   |
   = note: expected reference `&[Arg]`
              found reference `&realstd::vec::Vec<ffi::os_str::OsString>`
error: aborting due to 2 previous errors

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `std`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "i686-pc-windows-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "--all-targets" "-p" "test" "-p" "panic_abort" "-p" "alloc" "-p" "proc_macro" "-p" "panic_unwind" "-p" "unwind" "-p" "std" "-p" "term" "-p" "core" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu --all-targets
Build completed unsuccessfully in 0:00:28
```

</details>

dgrunwald commented:
I haven't tried this branch; but if I read the diff correctly, `unquoted_arg` will still perform escaping?
`Quote::Never` doesn't add quotes around the argument, but it still escapes those within the argument.

You don't have any test cases where escaping would trigger (only when the argument contains `"`).

dgrunwald commented:
Also, isn't there a bug where `Quote::Auto` ends up discarding trailing backslashes from the argument if the argument doesn't contain any whitespace?

kornelski commented:
There is a test verifying that there's no escaping applied to the unquoted args:

https://github.com/rust-lang/rust/pull/85832/files#diff-a426d649b902b742a112ec5836a040041fc4ba3dd4abc9a3927a901df46f2e36R12

kornelski commented:
I've disabled inner escaping for unquoted arg.

I haven't changed escaping logic. I think it's fine — the rules of CommandLineToArgvW that it uses are just very counter-intuitive.

dgrunwald commented:
Try a test case with `Arg::Quoted(OsString::from("c:\\")),` (without force_quotes).
I think it'll end up passing `c:` without the trailing backslash to the target program. That's because `backslashes` is non-zero at the end, but gets ignored when the `if quote` is not taken.

yaahc commented:
Why'd you go with calling it `quoted_arg` instead of `raw_arg` as suggested in the source issue?

kornelski commented:
That's because I've noticed that there's `force_quotes`, and wanted to use a similar teminology. However, you're right that this makes escaping ambiguous. I've changed back to `raw_arg`

rivy commented on library/std/src/sys/windows/process.rs:
```diff
@@ -97,6 +97,14 @@ pub struct StdioPipes {
+#[derive(Debug)]
+enum Arg {
+    /// Add quotes (if needed)
+    Regular(OsString),
```
```suggestion
    AutoQuote(OsString),
```

maybe? It adds semantic meaning.

kornelski replied:
I think it's fine as-is. It's for the regular `arg`, not `arg_auto` or such.

yaahc commented:
> Try a test case with `Arg::Quoted(OsString::from("c:\\")),` (without force_quotes).
> I think it'll end up passing `c:` without the trailing backslash to the target program. That's because `backslashes` is non-zero at the end, but gets ignored when the `if quote` is not taken.

@kornelski have you tried the test case suggested by @dgrunwald, I'm curious to know if this is an issue.

bors commented:
:umbrella: The latest upstream changes (presumably #85270) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ChrisDenton commented on library/std/src/os/windows/process.rs:
```diff
@@ -125,6 +126,13 @@ pub trait CommandExt: Sealed {
+    /// This is useful for passing arguments to `cmd.exe /c`, which doesn't follow
+    /// `CommandLineToArgvW` escaping rules.
+    #[unstable(feature = "windows_process_extensions_raw_arg", issue = "29494")]
+    fn raw_arg(&mut self, text_to_append_as_is: &OsStr) -> &mut process::Command;
```
The signature of [`Command::arg`](https://doc.rust-lang.org/std/process/struct.Command.html#method.arg) uses the generic `AsRef<OsStr>` for the type. This should probably match that?

kornelski replied:
I've added it. I'm not sure if `CommandExt` was supposed to be object-safe tho.

kornelski commented:
@yaahc I've added tests cases for the issue raised. They're fine, because escaping rules on Windows don't double the number of backslashes, only number of backslashes before a quote (the syntax is illogical like that by design).

ChrisDenton commented:
To recap, a backslash is treated literally unless they are directly followed by a double quote. When a double quote is encountered the rules are:
* Any directly preceding backslashes are counted (represented here by `n`).
    * If `n` is even then replace with `n/2` backslashes and toggle "quote mode".
    * If `n` is odd then replace with `(n-1)/2` backslashes and add a literal `"`.

For example, the string `C:\Program Files\Rust` can be encoded in an argument in several ways:

`"C:\Program Files\Rust"`
`C:\"Program Files"\Rust`
`C:\"Program Files\\"Rust`

These rules are based on the old 80's DOS shell and can't be changed too much due to backwards compatibility. However, some applications do use their own rules, hence the need for `raw_arg`.

dgrunwald commented:
I misread the code; I missed that the first copy of the `\` is already inserted by the loop.
So the code works fine, but thanks for adding the test.

Yeah the rules are weird. Fun fact: the interpretation of double double-quotes `""` is another (undocumented) special case where the behavior changed in the 2008 update of the Microsoft C runtime. But that doesn't matter for us as we never emit those. This site has the full story: http://daviddeley.com/autohotkey/parameters/parameters.htm#WINCRULES

yaahc commented:
Sounds like everyone is happy with the PR in it's current state ^_^

@bors r+

bors commented:
:pushpin: Commit b86d121c3af1d82425074ae8c005fa96b824aa1d has been approved by `yaahc`

<!-- @bors r=yaahc b86d121c3af1d82425074ae8c005fa96b824aa1d -->
<!-- homu: {"type":"Approved","sha":"b86d121c3af1d82425074ae8c005fa96b824aa1d","approver":"yaahc"} -->

bors commented:
:hourglass: Testing commit b86d121c3af1d82425074ae8c005fa96b824aa1d with merge f51c808dad50ff0d0849bd4becacf35d3728f8a6...
<!-- homu: {"type":"BuildStarted","head_sha":"b86d121c3af1d82425074ae8c005fa96b824aa1d","merge_sha":"f51c808dad50ff0d0849bd4becacf35d3728f8a6"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

ChrisDenton commented:
Is that just homu acting up?

kornelski commented:
@yaahc bors needs a retry

yaahc commented:
@bors retry

bors commented:
:hourglass: Testing commit b86d121c3af1d82425074ae8c005fa96b824aa1d with merge 61e56dc608b0077df1cdc469c0eea3e9a72275ce...
<!-- homu: {"type":"BuildStarted","head_sha":"b86d121c3af1d82425074ae8c005fa96b824aa1d","merge_sha":"61e56dc608b0077df1cdc469c0eea3e9a72275ce"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3024760589)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/3024760589","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`i686-mingw-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/3024431406) [(plain)](https://github.com/rust-lang-ci/rust/commit/61e56dc608b0077df1cdc469c0eea3e9a72275ce/checks/3024431406/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
status: exit code: 2
command: "make"
stdout:
------------------------------------------
make[1]: Entering directory '/d/a/rust/rust/src/test/run-make-fulldeps/link-args-order'
PATH="/d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-args-order/link-args-order:D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0-bootstrap-tools/i686-pc-windows-gnu/release/deps:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0/bin:/d/a/rust/rust/ninja:/d/a/rust/rust/mingw32/bin:/c/hostedtoolcache/windows/Python/3.9.5/x64/Scripts:/c/hostedtoolcache/windows/Python/3.9.5/x64:/usr/bin:/d/a/rust/rust/sccache:/c/Users/runneradmin/.dotnet/tools:/c/Program Files/MongoDB/Server/4.4/bin:/c/aliyun-cli:/c/vcpkg:/c/cf-cli:/c/Program Files (x86)/NSIS:/c/tools/zstd:/c/Program Files/Mercurial:/c/hostedtoolcache/windows/stack/2.7.1/x64:/c/tools/ghc-9.0.1/bin:/c/Program Files/dotnet:/c/mysql-5.7.21-winx64/bin:/c/Program Files/R/R-4.1.0/bin/x64:/c/SeleniumWebDrivers/GeckoDriver:/c/Program Files (x86)/sbt/bin:/c/Rust/.cargo/bin:/c/Program Files (x86)/GitHub CLI:/c/Program Files/Git/bin:/c/Program Files (x86)/pipx_bin:/c/hostedtoolcache/windows/go/1.15.13/x64/bin:/c/hostedtoolcache/windows/Python/3.7.9/x64/Scripts:/c/hostedtoolcache/windows/Python/3.7.9/x64:/c/hostedtoolcache/windows/Ruby/2.5.9/x64/bin:/c/hostedtoolcache/windows/Java_Adopt_jdk/8.0.292-10/x64/bin:/c/npm/prefix:/c/Program Files (x86)/Microsoft SDKs/Azure/CLI2/wbin:/c/ProgramData/kind:/c/Windows/system32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/c/Windows/System32/OpenSSH:/c/ProgramData/Chocolatey/bin:/c/Program Files/Microsoft/Web Platform Installer:/c/Program Files/Docker:/c/Program Files/PowerShell/7:/c/Program Files/dotnet:/c/Program Files/Microsoft SQL Server/130/Tools/Binn:/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/170/Tools/Binn:/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit:/c/Program Files (x86)/Microsoft SQL Server/110/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/130/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/140/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/150/DTS/Binn:/c/Program Files/nodejs:/c/ProgramData/chocolatey/lib/pulumi/tools/Pulumi/bin:/c/ProgramData/chocolatey/lib/maven/apache-maven-3.8.1/bin:/c/Program Files/Microsoft Service Fabric/bin/Fabric/Fabric.Code:/c/Program Files/Microsoft SDKs/Service Fabric/Tools/ServiceFabricLocalClusterManager:/c/Program Files/OpenSSL/bin:/c/Strawberry/c/bin:/c/Strawberry/perl/site/bin:/c/Strawberry/perl/bin:/c/Program Files/Git/cmd:/c/Program Files/Git/mingw64/bin:/c/Program Files/Git/usr/bin:/c/tools/php:/c/Program Files (x86)/sbt/bin:/c/Program Files/TortoiseSVN/bin:/c/SeleniumWebDrivers/ChromeDriver:/c/SeleniumWebDrivers/EdgeDriver:/c/Program Files/CMake/bin:/c/Program Files/Amazon/AWSCLIV2:/c/Program Files/Amazon/SessionManagerPlugin/bin:/c/Program Files/Amazon/AWSSAMCLI/bin:/c/Program Files (x86)/Google/Cloud SDK/google-cloud-sdk/bin:/c/Program Files (x86)/Microsoft BizTalk Server:/c/Users/runneradmin/AppData/Local/Microsoft/WindowsApps" 'D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin\rustc.exe' --out-dir /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-args-order/link-args-order -L /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-args-order/link-args-order  -C linker-flavor=ld -C link-arg=a -C link-args="b c" -C link-args="d e" -C link-arg=f empty.rs 2>&1 | "D:\a\rust\rust/src/etc/cat-and-grep.sh" '"a" "b" "c" "d" "e" "f"'
[[[ begin stdout ]]]
error: linking with `ld` failed: exit code: 1
  |
  = note: "ld" Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\rsbegin.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-args-order/link-args-order\\empty.empty.7ea5ba47-cgu.0.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-args-order/link-args-order\\empty.empty.7ea5ba47-cgu.1.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-args-order/link-args-order\\empty.empty.7ea5ba47-cgu.2.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-args-order/link-args-order\\empty.empty.7ea5ba47-cgu.3.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-args-order/link-args-order\\empty.empty.7ea5ba47-cgu.4.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-args-order/link-args-order\\empty.empty.7ea5ba47-cgu.5.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-args-order/link-args-order\\empty.empty.7ea5ba47-cgu.6.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-args-order/link-args-order\\empty.empty.7ea5ba47-cgu.7.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-args-order/link-args-order\\empty.empty.7ea5ba47-cgu.8.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-args-order/link-args-order\\empty.57fr01r08et8q3u0.rcgu.o") Regular("-L") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-args-order/link-args-order") Regular("-L") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib") Regular("--start-group") Regular("-Bstatic") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libstd-8a141fb1ca16a7f4.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libpanic_unwind-b4dc0e5ec3617f0e.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libobject-e3dc7427007d0937.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libaddr2line-b5a829ff7543314a.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libgimli-c425bc7c78ddb7d3.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libstd_detect-366a35e39f251bff.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\librustc_demangle-a7f7f36224101ef9.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libhashbrown-4313d8aeaeb625c1.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\librustc_std_workspace_alloc-e13c460dcc369e92.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libunwind-f951f52262eb1201.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libcfg_if-2aa9b9bc5372666d.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\liblibc-e2d0cc833bfd62f3.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\liballoc-0c35ae82faa342f3.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\librustc_std_workspace_core-d4ca5ea87aa7e9c1.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libcore-ce3832ec38af8636.rlib") Regular("--end-group") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libcompiler_builtins-a84d35280d75ab62.rlib") Regular("-Bdynamic") Regular("-lkernel32") Regular("-lws2_32") Regular("-ladvapi32") Regular("-luserenv") Regular("-lkernel32") Regular("--nxcompat") Regular("-L") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib") Regular("-o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-args-order/link-args-order\\empty.exe") Regular("--gc-sections") Regular("a") Regular("b") Regular("c") Regular("d") Regular("e") Regular("f") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\rsend.o")
Some tests failed in compiletest suite=run-make-fulldeps mode=run-make host=i686-pc-windows-gnu target=i686-pc-windows-gnu
  = note: D:\a\rust\rust\mingw32\bin\ld.exe: cannot find -lkernel32
          D:\a\rust\rust\mingw32\bin\ld.exe: cannot find -lws2_32
          D:\a\rust\rust\mingw32\bin\ld.exe: cannot find -ladvapi32
          D:\a\rust\rust\mingw32\bin\ld.exe: cannot find -luserenv
          D:\a\rust\rust\mingw32\bin\ld.exe: cannot find -lkernel32
          D:\a\rust\rust\mingw32\bin\ld.exe: cannot find a: No such file or directory
          D:\a\rust\rust\mingw32\bin\ld.exe: cannot find b: No such file or directory
          D:\a\rust\rust\mingw32\bin\ld.exe: cannot find c: No such file or directory
          D:\a\rust\rust\mingw32\bin\ld.exe: cannot find d: No such file or directory
          D:\a\rust\rust\mingw32\bin\ld.exe: cannot find e: No such file or directory
          D:\a\rust\rust\mingw32\bin\ld.exe: cannot find f: No such file or directory

error: aborting due to previous error



[[[ end stdout ]]]
Error: cannot match: "a" "b" "c" "d" "e" "f"
make[1]: Leaving directory '/d/a/rust/rust/src/test/run-make-fulldeps/link-args-order'
------------------------------------------
stderr:
------------------------------------------
------------------------------------------
make[1]: *** [Makefile:9: all] Error 1
------------------------------------------


---- [run-make] run-make-fulldeps\link-dedup stdout ----
---- [run-make] run-make-fulldeps\link-dedup stdout ----

error: make failed
status: exit code: 2
command: "make"
stdout:
------------------------------------------
make[1]: Entering directory '/d/a/rust/rust/src/test/run-make-fulldeps/link-dedup'
PATH="/d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup:D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0-bootstrap-tools/i686-pc-windows-gnu/release/deps:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0/bin:/d/a/rust/rust/ninja:/d/a/rust/rust/mingw32/bin:/c/hostedtoolcache/windows/Python/3.9.5/x64/Scripts:/c/hostedtoolcache/windows/Python/3.9.5/x64:/usr/bin:/d/a/rust/rust/sccache:/c/Users/runneradmin/.dotnet/tools:/c/Program Files/MongoDB/Server/4.4/bin:/c/aliyun-cli:/c/vcpkg:/c/cf-cli:/c/Program Files (x86)/NSIS:/c/tools/zstd:/c/Program Files/Mercurial:/c/hostedtoolcache/windows/stack/2.7.1/x64:/c/tools/ghc-9.0.1/bin:/c/Program Files/dotnet:/c/mysql-5.7.21-winx64/bin:/c/Program Files/R/R-4.1.0/bin/x64:/c/SeleniumWebDrivers/GeckoDriver:/c/Program Files (x86)/sbt/bin:/c/Rust/.cargo/bin:/c/Program Files (x86)/GitHub CLI:/c/Program Files/Git/bin:/c/Program Files (x86)/pipx_bin:/c/hostedtoolcache/windows/go/1.15.13/x64/bin:/c/hostedtoolcache/windows/Python/3.7.9/x64/Scripts:/c/hostedtoolcache/windows/Python/3.7.9/x64:/c/hostedtoolcache/windows/Ruby/2.5.9/x64/bin:/c/hostedtoolcache/windows/Java_Adopt_jdk/8.0.292-10/x64/bin:/c/npm/prefix:/c/Program Files (x86)/Microsoft SDKs/Azure/CLI2/wbin:/c/ProgramData/kind:/c/Windows/system32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/c/Windows/System32/OpenSSH:/c/ProgramData/Chocolatey/bin:/c/Program Files/Microsoft/Web Platform Installer:/c/Program Files/Docker:/c/Program Files/PowerShell/7:/c/Program Files/dotnet:/c/Program Files/Microsoft SQL Server/130/Tools/Binn:/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/170/Tools/Binn:/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit:/c/Program Files (x86)/Microsoft SQL Server/110/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/130/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/140/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/150/DTS/Binn:/c/Program Files/nodejs:/c/ProgramData/chocolatey/lib/pulumi/tools/Pulumi/bin:/c/ProgramData/chocolatey/lib/maven/apache-maven-3.8.1/bin:/c/Program Files/Microsoft Service Fabric/bin/Fabric/Fabric.Code:/c/Program Files/Microsoft SDKs/Service Fabric/Tools/ServiceFabricLocalClusterManager:/c/Program Files/OpenSSL/bin:/c/Strawberry/c/bin:/c/Strawberry/perl/site/bin:/c/Strawberry/perl/bin:/c/Program Files/Git/cmd:/c/Program Files/Git/mingw64/bin:/c/Program Files/Git/usr/bin:/c/tools/php:/c/Program Files (x86)/sbt/bin:/c/Program Files/TortoiseSVN/bin:/c/SeleniumWebDrivers/ChromeDriver:/c/SeleniumWebDrivers/EdgeDriver:/c/Program Files/CMake/bin:/c/Program Files/Amazon/AWSCLIV2:/c/Program Files/Amazon/SessionManagerPlugin/bin:/c/Program Files/Amazon/AWSSAMCLI/bin:/c/Program Files (x86)/Google/Cloud SDK/google-cloud-sdk/bin:/c/Program Files (x86)/Microsoft BizTalk Server:/c/Users/runneradmin/AppData/Local/Microsoft/WindowsApps" 'D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin\rustc.exe' --out-dir /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup -L /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup  depa.rs
PATH="/d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup:D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0-bootstrap-tools/i686-pc-windows-gnu/release/deps:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0/bin:/d/a/rust/rust/ninja:/d/a/rust/rust/mingw32/bin:/c/hostedtoolcache/windows/Python/3.9.5/x64/Scripts:/c/hostedtoolcache/windows/Python/3.9.5/x64:/usr/bin:/d/a/rust/rust/sccache:/c/Users/runneradmin/.dotnet/tools:/c/Program Files/MongoDB/Server/4.4/bin:/c/aliyun-cli:/c/vcpkg:/c/cf-cli:/c/Program Files (x86)/NSIS:/c/tools/zstd:/c/Program Files/Mercurial:/c/hostedtoolcache/windows/stack/2.7.1/x64:/c/tools/ghc-9.0.1/bin:/c/Program Files/dotnet:/c/mysql-5.7.21-winx64/bin:/c/Program Files/R/R-4.1.0/bin/x64:/c/SeleniumWebDrivers/GeckoDriver:/c/Program Files (x86)/sbt/bin:/c/Rust/.cargo/bin:/c/Program Files (x86)/GitHub CLI:/c/Program Files/Git/bin:/c/Program Files (x86)/pipx_bin:/c/hostedtoolcache/windows/go/1.15.13/x64/bin:/c/hostedtoolcache/windows/Python/3.7.9/x64/Scripts:/c/hostedtoolcache/windows/Python/3.7.9/x64:/c/hostedtoolcache/windows/Ruby/2.5.9/x64/bin:/c/hostedtoolcache/windows/Java_Adopt_jdk/8.0.292-10/x64/bin:/c/npm/prefix:/c/Program Files (x86)/Microsoft SDKs/Azure/CLI2/wbin:/c/ProgramData/kind:/c/Windows/system32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/c/Windows/System32/OpenSSH:/c/ProgramData/Chocolatey/bin:/c/Program Files/Microsoft/Web Platform Installer:/c/Program Files/Docker:/c/Program Files/PowerShell/7:/c/Program Files/dotnet:/c/Program Files/Microsoft SQL Server/130/Tools/Binn:/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/170/Tools/Binn:/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit:/c/Program Files (x86)/Microsoft SQL Server/110/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/130/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/140/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/150/DTS/Binn:/c/Program Files/nodejs:/c/ProgramData/chocolatey/lib/pulumi/tools/Pulumi/bin:/c/ProgramData/chocolatey/lib/maven/apache-maven-3.8.1/bin:/c/Program Files/Microsoft Service Fabric/bin/Fabric/Fabric.Code:/c/Program Files/Microsoft SDKs/Service Fabric/Tools/ServiceFabricLocalClusterManager:/c/Program Files/OpenSSL/bin:/c/Strawberry/c/bin:/c/Strawberry/perl/site/bin:/c/Strawberry/perl/bin:/c/Program Files/Git/cmd:/c/Program Files/Git/mingw64/bin:/c/Program Files/Git/usr/bin:/c/tools/php:/c/Program Files (x86)/sbt/bin:/c/Program Files/TortoiseSVN/bin:/c/SeleniumWebDrivers/ChromeDriver:/c/SeleniumWebDrivers/EdgeDriver:/c/Program Files/CMake/bin:/c/Program Files/Amazon/AWSCLIV2:/c/Program Files/Amazon/SessionManagerPlugin/bin:/c/Program Files/Amazon/AWSSAMCLI/bin:/c/Program Files (x86)/Google/Cloud SDK/google-cloud-sdk/bin:/c/Program Files (x86)/Microsoft BizTalk Server:/c/Users/runneradmin/AppData/Local/Microsoft/WindowsApps" 'D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin\rustc.exe' --out-dir /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup -L /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup  depb.rs
PATH="/d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup:D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0-bootstrap-tools/i686-pc-windows-gnu/release/deps:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0/bin:/d/a/rust/rust/ninja:/d/a/rust/rust/mingw32/bin:/c/hostedtoolcache/windows/Python/3.9.5/x64/Scripts:/c/hostedtoolcache/windows/Python/3.9.5/x64:/usr/bin:/d/a/rust/rust/sccache:/c/Users/runneradmin/.dotnet/tools:/c/Program Files/MongoDB/Server/4.4/bin:/c/aliyun-cli:/c/vcpkg:/c/cf-cli:/c/Program Files (x86)/NSIS:/c/tools/zstd:/c/Program Files/Mercurial:/c/hostedtoolcache/windows/stack/2.7.1/x64:/c/tools/ghc-9.0.1/bin:/c/Program Files/dotnet:/c/mysql-5.7.21-winx64/bin:/c/Program Files/R/R-4.1.0/bin/x64:/c/SeleniumWebDrivers/GeckoDriver:/c/Program Files (x86)/sbt/bin:/c/Rust/.cargo/bin:/c/Program Files (x86)/GitHub CLI:/c/Program Files/Git/bin:/c/Program Files (x86)/pipx_bin:/c/hostedtoolcache/windows/go/1.15.13/x64/bin:/c/hostedtoolcache/windows/Python/3.7.9/x64/Scripts:/c/hostedtoolcache/windows/Python/3.7.9/x64:/c/hostedtoolcache/windows/Ruby/2.5.9/x64/bin:/c/hostedtoolcache/windows/Java_Adopt_jdk/8.0.292-10/x64/bin:/c/npm/prefix:/c/Program Files (x86)/Microsoft SDKs/Azure/CLI2/wbin:/c/ProgramData/kind:/c/Windows/system32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/c/Windows/System32/OpenSSH:/c/ProgramData/Chocolatey/bin:/c/Program Files/Microsoft/Web Platform Installer:/c/Program Files/Docker:/c/Program Files/PowerShell/7:/c/Program Files/dotnet:/c/Program Files/Microsoft SQL Server/130/Tools/Binn:/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/170/Tools/Binn:/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit:/c/Program Files (x86)/Microsoft SQL Server/110/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/130/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/140/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/150/DTS/Binn:/c/Program Files/nodejs:/c/ProgramData/chocolatey/lib/pulumi/tools/Pulumi/bin:/c/ProgramData/chocolatey/lib/maven/apache-maven-3.8.1/bin:/c/Program Files/Microsoft Service Fabric/bin/Fabric/Fabric.Code:/c/Program Files/Microsoft SDKs/Service Fabric/Tools/ServiceFabricLocalClusterManager:/c/Program Files/OpenSSL/bin:/c/Strawberry/c/bin:/c/Strawberry/perl/site/bin:/c/Strawberry/perl/bin:/c/Program Files/Git/cmd:/c/Program Files/Git/mingw64/bin:/c/Program Files/Git/usr/bin:/c/tools/php:/c/Program Files (x86)/sbt/bin:/c/Program Files/TortoiseSVN/bin:/c/SeleniumWebDrivers/ChromeDriver:/c/SeleniumWebDrivers/EdgeDriver:/c/Program Files/CMake/bin:/c/Program Files/Amazon/AWSCLIV2:/c/Program Files/Amazon/SessionManagerPlugin/bin:/c/Program Files/Amazon/AWSSAMCLI/bin:/c/Program Files (x86)/Google/Cloud SDK/google-cloud-sdk/bin:/c/Program Files (x86)/Microsoft BizTalk Server:/c/Users/runneradmin/AppData/Local/Microsoft/WindowsApps" 'D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin\rustc.exe' --out-dir /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup -L /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup  depc.rs
PATH="/d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup:D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0-bootstrap-tools/i686-pc-windows-gnu/release/deps:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0/bin:/d/a/rust/rust/ninja:/d/a/rust/rust/mingw32/bin:/c/hostedtoolcache/windows/Python/3.9.5/x64/Scripts:/c/hostedtoolcache/windows/Python/3.9.5/x64:/usr/bin:/d/a/rust/rust/sccache:/c/Users/runneradmin/.dotnet/tools:/c/Program Files/MongoDB/Server/4.4/bin:/c/aliyun-cli:/c/vcpkg:/c/cf-cli:/c/Program Files (x86)/NSIS:/c/tools/zstd:/c/Program Files/Mercurial:/c/hostedtoolcache/windows/stack/2.7.1/x64:/c/tools/ghc-9.0.1/bin:/c/Program Files/dotnet:/c/mysql-5.7.21-winx64/bin:/c/Program Files/R/R-4.1.0/bin/x64:/c/SeleniumWebDrivers/GeckoDriver:/c/Program Files (x86)/sbt/bin:/c/Rust/.cargo/bin:/c/Program Files (x86)/GitHub CLI:/c/Program Files/Git/bin:/c/Program Files (x86)/pipx_bin:/c/hostedtoolcache/windows/go/1.15.13/x64/bin:/c/hostedtoolcache/windows/Python/3.7.9/x64/Scripts:/c/hostedtoolcache/windows/Python/3.7.9/x64:/c/hostedtoolcache/windows/Ruby/2.5.9/x64/bin:/c/hostedtoolcache/windows/Java_Adopt_jdk/8.0.292-10/x64/bin:/c/npm/prefix:/c/Program Files (x86)/Microsoft SDKs/Azure/CLI2/wbin:/c/ProgramData/kind:/c/Windows/system32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/c/Windows/System32/OpenSSH:/c/ProgramData/Chocolatey/bin:/c/Program Files/Microsoft/Web Platform Installer:/c/Program Files/Docker:/c/Program Files/PowerShell/7:/c/Program Files/dotnet:/c/Program Files/Microsoft SQL Server/130/Tools/Binn:/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/170/Tools/Binn:/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit:/c/Program Files (x86)/Microsoft SQL Server/110/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/130/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/140/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/150/DTS/Binn:/c/Program Files/nodejs:/c/ProgramData/chocolatey/lib/pulumi/tools/Pulumi/bin:/c/ProgramData/chocolatey/lib/maven/apache-maven-3.8.1/bin:/c/Program Files/Microsoft Service Fabric/bin/Fabric/Fabric.Code:/c/Program Files/Microsoft SDKs/Service Fabric/Tools/ServiceFabricLocalClusterManager:/c/Program Files/OpenSSL/bin:/c/Strawberry/c/bin:/c/Strawberry/perl/site/bin:/c/Strawberry/perl/bin:/c/Program Files/Git/cmd:/c/Program Files/Git/mingw64/bin:/c/Program Files/Git/usr/bin:/c/tools/php:/c/Program Files (x86)/sbt/bin:/c/Program Files/TortoiseSVN/bin:/c/SeleniumWebDrivers/ChromeDriver:/c/SeleniumWebDrivers/EdgeDriver:/c/Program Files/CMake/bin:/c/Program Files/Amazon/AWSCLIV2:/c/Program Files/Amazon/SessionManagerPlugin/bin:/c/Program Files/Amazon/AWSSAMCLI/bin:/c/Program Files (x86)/Google/Cloud SDK/google-cloud-sdk/bin:/c/Program Files (x86)/Microsoft BizTalk Server:/c/Users/runneradmin/AppData/Local/Microsoft/WindowsApps" 'D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin\rustc.exe' --out-dir /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup -L /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup  empty.rs --cfg bar 2>&1 | "D:\a\rust\rust/src/etc/cat-and-grep.sh" '"-ltesta" "-ltestb" "-ltesta"'
[[[ begin stdout ]]]
error: linking with `i686-w64-mingw32-gcc` failed: exit code: 1
  |
  = note: "i686-w64-mingw32-gcc" Regular("-fno-use-linker-plugin") Regular("-Wl,--dynamicbase") Regular("-Wl,--disable-auto-image-base") Regular("-Wl,--large-address-aware") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\rsbegin.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup\\empty.empty.7ea5ba47-cgu.0.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup\\empty.empty.7ea5ba47-cgu.1.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup\\empty.empty.7ea5ba47-cgu.2.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup\\empty.empty.7ea5ba47-cgu.3.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup\\empty.empty.7ea5ba47-cgu.4.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup\\empty.empty.7ea5ba47-cgu.5.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup\\empty.empty.7ea5ba47-cgu.6.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup\\empty.empty.7ea5ba47-cgu.7.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup\\empty.empty.7ea5ba47-cgu.8.rcgu.o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup\\empty.57fr01r08et8q3u0.rcgu.o") Regular("-L") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup") Regular("-L") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib") Regular("-Wl,-Bstatic") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\test\\run-make-fulldeps\\link-dedup\\link-dedup\\libdepc.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\test\\run-make-fulldeps\\link-dedup\\link-dedup\\libdepb.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\test\\run-make-fulldeps\\link-dedup\\link-dedup\\libdepa.rlib") Regular("-Wl,--start-group") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libstd-8a141fb1ca16a7f4.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libpanic_unwind-b4dc0e5ec3617f0e.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libobject-e3dc7427007d0937.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libaddr2line-b5a829ff7543314a.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libgimli-c425bc7c78ddb7d3.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libstd_detect-366a35e39f251bff.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\librustc_demangle-a7f7f36224101ef9.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libhashbrown-4313d8aeaeb625c1.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\librustc_std_workspace_alloc-e13c460dcc369e92.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libunwind-f951f52262eb1201.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libcfg_if-2aa9b9bc5372666d.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\liblibc-e2d0cc833bfd62f3.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\liballoc-0c35ae82faa342f3.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\librustc_std_workspace_core-d4ca5ea87aa7e9c1.rlib") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libcore-ce3832ec38af8636.rlib") Regular("-Wl,--end-group") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\libcompiler_builtins-a84d35280d75ab62.rlib") Regular("-Wl,-Bdynamic") Regular("-ltesta") Regular("-ltestb") Regular("-ltesta") Regular("-lkernel32") Regular("-lws2_32") Regular("-ladvapi32") Regular("-luserenv") Regular("-lkernel32") Regular("-lgcc_eh") Regular("-l:libpthread.a") Regular("-lmsvcrt") Regular("-lmingwex") Regular("-lmingw32") Regular("-lgcc") Regular("-lmsvcrt") Regular("-luser32") Regular("-lkernel32") Regular("-Wl,--nxcompat") Regular("-L") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib") Regular("-o") Regular("D:/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/link-dedup/link-dedup\\empty.exe") Regular("-Wl,--gc-sections") Regular("-no-pie") Regular("-nodefaultlibs") Regular("D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\rsend.o")
  = note: D:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/6.3.0/../../../../i686-w64-mingw32/bin/ld.exe: cannot find -ltesta
          D:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/6.3.0/../../../../i686-w64-mingw32/bin/ld.exe: cannot find -ltestb
          D:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/6.3.0/../../../../i686-w64-mingw32/bin/ld.exe: cannot find -ltesta
          collect2.exe: error: ld returned 1 exit status

error: aborting due to previous error



[[[ end stdout ]]]
Error: cannot match: "-ltesta" "-ltestb" "-ltesta"
make[1]: Leaving directory '/d/a/rust/rust/src/test/run-make-fulldeps/link-dedup'
------------------------------------------
stderr:
------------------------------------------
------------------------------------------
make[1]: *** [Makefile:9: all] Error 1
------------------------------------------



---
test result: FAILED. 183 passed; 2 failed; 39 ignored; 0 measured; 0 filtered out; finished in 99.61s



command did not execute successfully: "D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage0-tools-bin\\compiletest.exe" "--compile-lib-path" "D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\bin" "--run-lib-path" "D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\lib\\rustlib\\i686-pc-windows-gnu\\lib" "--rustc-path" "D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\bin\\rustc.exe" "--rustdoc-path" "D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\bin\\rustdoc.exe" "--rust-demangler-path" "D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2-tools-bin\\rust-demangler.exe" "--src-base" "D:\\a\\rust\\rust\\src/test\\run-make-fulldeps" "--build-base" "D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\test\\run-make-fulldeps" "--stage-id" "stage2-i686-pc-windows-gnu" "--suite" "run-make-fulldeps" "--mode" "run-make" "--target" "i686-pc-windows-gnu" "--host" "i686-pc-windows-gnu" "--llvm-filecheck" "D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\llvm\\build\\bin\\FileCheck.exe" "--nodejs" "C:\\Program Files\\nodejs\\node" "--npm" "C:\\Program Files\\nodejs\\npm" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\native\\rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\native\\rust-test-helpers" "--docck-python" "C:\\hostedtoolcache\\windows\\Python\\3.9.5\\x64\\python3.exe" "--lldb-python" "C:\\hostedtoolcache\\windows\\Python\\3.9.5\\x64\\python3.exe" "--gdb" "D:\\a\\rust\\rust\\mingw32\\bin\\gdb" "--llvm-version" "12.0.1-rust-1.55.0-nightly" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xray" "--llvm-bin-dir" "D:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\llvm\\build\\bin" "--cc" "gcc.exe" "--cxx" "g++.exe" "--cflags" "-ffunction-sections -fdata-sections -m32 -fno-omit-frame-pointer" "--ar" "ar" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"



make: *** [Makefile:80: ci-mingw-subset-1] Error 1
```

</details>

kornelski commented:
I think mingw uses debug output in tests? I've changed debug output to look exactly like before, and rebased.  

I can't test mingw, because I don't know how to make x.py use it instead of msvc.

yaahc commented:
I'm not really familiar with windows at all so I can't really help with the mingw/msvc setup. For now though I can delegate bors permissions so you can restart the build as needed

@bors delegate+
@bors r+

bors commented:
:v: @kornelski can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"yaahc","delegate":"kornelski"} -->

bors commented:
:pushpin: Commit bc67f6bc95106ef0e0b4681a384af153a7228bf1 has been approved by `yaahc`

<!-- @bors r=yaahc bc67f6bc95106ef0e0b4681a384af153a7228bf1 -->
<!-- homu: {"type":"Approved","sha":"bc67f6bc95106ef0e0b4681a384af153a7228bf1","approver":"yaahc"} -->

bors commented:
:hourglass: Testing commit bc67f6bc95106ef0e0b4681a384af153a7228bf1 with merge 3eff244fc7340c14b34299c1ff70ec1a76202332...
<!-- homu: {"type":"BuildStarted","head_sha":"bc67f6bc95106ef0e0b4681a384af153a7228bf1","merge_sha":"3eff244fc7340c14b34299c1ff70ec1a76202332"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3032663624)
Approved by: yaahc
Pushing 3eff244fc7340c14b34299c1ff70ec1a76202332 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"yaahc","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3032663624"},"merge_sha":"3eff244fc7340c14b34299c1ff70ec1a76202332"} -->

Xaeroxe commented:
What's the path for stabilizing this feature? The unstable attribute for it doesn't point at a tracking issue.

kornelski commented:
@Xaeroxe There haven't been any objections to this feature, so I think you could make a tracking issue and make a PR to stabilize this feature.

Xaeroxe commented:
Tracking issue: https://github.com/rust-lang/rust/issues/92939
